### PR TITLE
Add NPU FusedAdam support

### PR DESCRIPTION
**Pull request by CurryRice233**
Add NPU FusedAdam support.

### CODE REVIEW

**Comment by CurryRice233** [comment]
@tjruwase @jeffra @RezaYazdaniAminabadi @cmikeh2 Sorry for annoying, can you guys review this PR ?

**Review by tjruwase — APPROVED** [review]

**Comment by statelesshz** [comment]
@tjruwase Good day. This PR is approved and ready to be merged. Could you retrigger this workflow and merge it? Thanks :-)

File: accelerator/npu_accelerator.py [hunk 1]
```diff
@@ -2,16 +2,27 @@
...
+try:
+    import torch.npu
+except ImportError:
+    pass
 
 
 class NPU_Accelerator(DeepSpeedAccelerator):
-
     def __init__(self):
+        super().__init__()
```
Comment (statelesshz) [inline accelerator/npu_accelerator.py:19]:
```suggestion

        super().__init__()
```

File: op_builder/npu/fused_adam.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+from .builder import NPUOpBuilder
+
+try:
+    import torch_npu
+except ImportError as e:
+    pass
+
+
+class NPUFusedAdam:
+    @staticmethod
```
Comment (statelesshz) [inline op_builder/npu/fused_adam.py:15]:
```suggestion

    @staticmethod
```

File: op_builder/npu/fused_adam.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    def multi_tensor_adam(chunk_size, noop_flag_buffer, tensor_lists, lr, beta1, beta2, epsilon, step, adam_w_mode,
+                          bias_correction, weight_decay, *args):
+        bias_correction1 = beta1 ** step
+        bias_correction2 = beta2 ** step
```
Comment (statelesshz) [inline op_builder/npu/fused_adam.py:19]:
```suggestion
        bias_correction2 = beta2**step
```

File: op_builder/npu/fused_adam.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    import torch_npu
+except ImportError as e:
+    pass
+
+
+class NPUFusedAdam:
+    @staticmethod
+    def multi_tensor_adam(chunk_size, noop_flag_buffer, tensor_lists, lr, beta1, beta2, epsilon, step, adam_w_mode,
+                          bias_correction, weight_decay, *args):
+        bias_correction1 = beta1 ** step
```
Comment (statelesshz) [inline op_builder/npu/fused_adam.py:18]:
```suggestion
        bias_correction1 = beta1**step
```

File: op_builder/npu/fused_adam.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                    weight_decay,
+                    beta1,
+                    beta2,
+                    epsilon,
+                    grad_flat,
+                    None,  # max_grad_norm
+                    False,  # amsgrad
+                    False,  # maximize
+                    out=(param_flat.data, m_flat, v_flat)
+                )
```
Comment (statelesshz) [inline op_builder/npu/fused_adam.py:42]:
```suggestion
```

File: op_builder/npu/fused_adam.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                    lr,
+                    weight_decay,
+                    beta1,
+                    beta2,
+                    epsilon,
+                    grad_flat,
+                    None,  # max_grad_norm
+                    False,  # amsgrad
+                    False,  # maximize
+                    out=(param_flat.data, m_flat, v_flat)
```
Comment (statelesshz) [inline op_builder/npu/fused_adam.py:41]:
```suggestion
                    out=(param_flat.data, m_flat, v_flat))
```

File: op_builder/npu/fused_adam.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                    bias_correction1,
+                    bias_correction2,
+                    lr,
+                    beta1,
+                    beta2,
+                    epsilon,
+                    grad_flat,
+                    False,  # use_locking
+                    False,  # use_nesterov
+                    out=(param_flat.data, m_flat, v_flat)
```
Comment (statelesshz) [inline op_builder/npu/fused_adam.py:54]:
```suggestion
                    out=(param_flat.data, m_flat, v_flat))
```

File: op_builder/npu/fused_adam.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                )
```
Comment (statelesshz) [inline op_builder/npu/fused_adam.py:55]:
```suggestion

```

**Review by statelesshz — COMMENTED** [review]
Resolve format checking errors

File: accelerator/npu_accelerator.py [hunk 9]
```diff
@@ -2,16 +2,28 @@
...
+try:
+    import torch.npu
+except ImportError:
+    pass
 
 
 class NPU_Accelerator(DeepSpeedAccelerator):
-
     def __init__(self):
+
```
Comment (statelesshz) [inline accelerator/npu_accelerator.py:19]:
```suggestion
```

File: accelerator/npu_accelerator.py [hunk 10]
```diff
@@ -2,16 +2,28 @@
...
+# allow op builder related API to be executed.
+try:
+    import torch.npu
+except ImportError:
+    pass
 
 
 class NPU_Accelerator(DeepSpeedAccelerator):
-
     def __init__(self):
```
Comment (statelesshz) [inline accelerator/npu_accelerator.py:18]:
```suggestion

    def __init__(self):
```

**Review by statelesshz — COMMENTED** [review]
As long as we modify these two blank lines, the format check error should be solved.

**Comment by tjruwase** [comment]
@CurryRice233, it is best to use this guide for formatting issues: https://github.com/microsoft/DeepSpeed/blob/master/CONTRIBUTING.md#prerequisites

**Comment by CurryRice233** [comment]
> @CurryRice233, it is best to use this guide for formatting issues: https://github.com/microsoft/DeepSpeed/blob/master/CONTRIBUTING.md#prerequisites

Thank you, new skill get😉.  By the way, could you retrigger this workflow again?

**Comment by CurryRice233** [comment]
@tjruwase hi, could you retrigger this workflow again and merge it? Thanks😀

**Pull request merged**